prep release: v1.48.0 (#5272)
apollographql/router: CHANGELOG.md

```yaml
telemetry:
  instrumentation:
    instruments:
      router:
        http.server.active_requests: false
        http.server.request.duration: false
        "custom_counter":
          description: "count of requests"
          type: counter
          unit: "unit"
          value: unit
          # This instrument should not be triggered as the condition is never true
          condition:
            eq:
              - response_header: "never-received"
              - static: "true"
```

In the case where a request was started, but the client aborted the request before the response was sent, the `response_header` would never be set to `"never-received"`,
and the instrument would not be triggered. However, the instrument would still be incremented.

Conditions are now checked for aborted requests, and the instrument is only incremented if the condition is fulfilled.

By [@BrynCooke](https://github.com/BrynCooke) in https://github.com/apollographql/router/pull/5215

## 🛠 Maintenance

### Send query planner and lifecycle metrics to Apollo ([PR #5267](https://github.com/apollographql/router/pull/5267), [PR #5270](https://github.com/apollographql/router/pull/5270))

> To allow us to measure how much of an improvement the new query planner implementation makes, we are now transmitting to Apollo new metrics:
> - `apollo.router.query_planning.*`  will greatly help us to improve query planning in the Router.
> - `apollo.router.lifecycle.api_schema` provides us feedback on the experimental Rust-based API schema generation.
> - `apollo.router.lifecycle.license` provides metrics on license expiration. We use this to improve the reliability of the license check mechanism.
> 
> These metrics do not leak any sensitive information.

Reviewer: ```suggestion
To enable the performance measurement of the router's new query planner implementation, the router transmits to Apollo the following new metrics:
- `apollo.router.query_planning.*` provides metrics on the query planner that help improve the query planning implementation.
- `apollo.router.lifecycle.api_schema` provides feedback on the experimental Rust-based API schema generation.
- `apollo.router.lifecycle.license` provides metrics on license expiration that help improve the reliability of the license check mechanism.

These metrics don't leak any sensitive data.
```

Reviewer: Re: `sensitive data`, is the type data sent by these metrics already of these  it covered in our [GraphOS data privacy docs](https://www.apollographql.com/docs/graphos/data-privacy)? If yes, can link it here, and if not, can add more context here.

Reviewer 2: I believe the thing @Geal is referring to is covered by the "metrics" that we already call out on this page, so we should be fine — https://www.apollographql.com/docs/graphos/data-privacy/#current-urls

By [@BrynCooke](https://github.com/BrynCooke) in https://github.com/apollographql/router/pull/5267, [@goto-bus-stop](https://github.com/goto-bus-stop)

## 🧪 Experimental

> ### Graphql instruments ([PR #5215](https://github.com/apollographql/router/pull/5215), [PR #5257](https://github.com/apollographql/router/pull/5257))

Reviewer: ```suggestion
### GraphQL instruments ([PR #5215](https://github.com/apollographql/router/pull/5215), [PR #5257](https://github.com/apollographql/router/pull/5257))
```

This PR adds experimental GraphQL instruments to telemetry.

> It makes the following possible:

Reviewer: ```suggestion
The new instruments are configured in the following:
```

```
telemetry:
  instrumentation:
    instruments:
      graphql:
        # The number of times a field was executed (counter)
        field.execution: true

        # The length of list fields (histogram)
        list.length: true

        # Custom counter of field execution where field name = name
        "custom_counter":
          description: "count of name field"
          type: counter
          unit: "unit"
          value: field_unit
          attributes:
            graphql.type.name: true
            graphql.field.type: true
            graphql.field.name: true
          condition:
            eq:
              - field_name: string
              - "name"

        # Custom histogram of list lengths for topProducts
        "custom_histogram":
          description: "histogram of review length"
          type: histogram
          unit: "unit"
          attributes:
            graphql.type.name: true
            graphql.field.type: true
            graphql.field.name: true
          value:
            field_custom:
              list_length: value
          condition:
            eq:
              - field_name: string
              - "topProducts"
```

> Note that this will have a significant performance impact which will be addressed in a following release.
> Users should also be aware that large numbers of excessive metrics may be generated, and they should take care not to run up a large APM bill.

Reviewer: ```suggestion
Using the new instruments consumes significant performance resources from the router. Their performance will be improved in a future release. 

Large numbers of metrics may also be generated by using the instruments, so make sure to not incur excessively large APM costs.
```

> For now, do not use these metrics in production.

Reviewer: ```suggestion
⚠ Use these instruments only in development. Don't use them in production.
```

# [1.48.0] - 2024-05-29

## 🚀 Features

### Add support for studio trace id in selectors and document context entry ([Issue #3803](https://github.com/apollographql/router/issues/3803)), ([Issue #5172](https://github.com/apollographql/router/issues/5172))

Add support for a new trace ID selector kind, the `apollo` trace ID, which represents the trace ID on [Apollo GraphOS Studio](https://studio.apollographql.com/). 

An example configuration using `trace_id: apollo`:

```yaml
telemetry:
  instrumentation:
    spans:
      router:
        "studio.trace.id":
            trace_id:: apollo
```

Also, router documentation has been updated with constants that are available in Rhai.

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/5189

### Add ability for router to deal with query plans with contextual rewrites ([PR #5097](https://github.com/apollographql/router/pull/5097))

Adds the ability for the router to execute query plans with context rewrites on them. These are generated by the @fromContext directive and are used to map a Value in the collected data JSON onto a variable which will in turn be used as an argument to a field resolver.

> ⚠️ This ships with a new version of federation, 2.8.0, which means distributed caches will be repopulated.

Reviewer 3: @clenfest - can we add a link to the docs here?

Author: I cannot find it in the router docs nor in the [PR](https://github.com/apollographql/router/commit/5a0826e6f22568bf61b99447501e2711bbf36edc) that introduced it. Could we add some docs for this feature before releasing it?

Author: @chandrikas the docs PR is in the federation repo https://github.com/apollographql/federation/pull/2975

By [@clenfest](https://github.com/clenfest) in https://github.com/apollographql/router/pull/5097

## 🐛 Fixes

> ### Submit new metrics to Apollo ([PR #5270](https://github.com/apollographql/router/pull/5270))

Reviewer 4: This isn't a fix. If we merge it, fine, but otherwise maybe it's a chore?

Reviewer 5: I put the other one under config, but we can move both?

The router submits two new metrics to Apollo:
- `apollo.router.lifecycle.api_schema` provides us feedback on the experimental Rust-based API schema generation.
- `apollo.router.lifecycle.license` provides metrics on license expiration. We use this to improve the reliability of the license check mechanism.

> 

Reviewer 6: perhaps we should merge this item with the query planning metrics one below

Previously when a telemetry instrument was dropped it would be incremented even if the associated condition was not fulfilled. For instance:

```yaml
telemetry:
  instrumentation:
    instruments:
      router:
        http.server.active_requests: false
        http.server.request.duration: false
        "custom_counter":
          description: "count of requests"
          type: counter
          unit: "unit"
          value: unit
          # This instrument should not be triggered as the condition is never true
          condition:
            eq:
              - response_header: "never-received"
              - static: "true"
```

In the case where a request was started, but the client aborted the request before the response was sent, the `response_header` would never be set to `"never-received"`,
and the instrument would not be triggered. However, the instrument would still be incremented.

Conditions are now checked for aborted requests, and the instrument is only incremented if the condition is fulfilled.

By [@BrynCooke](https://github.com/BrynCooke) in https://github.com/apollographql/router/pull/5215

## 📃 Configuration

> ### Sending query planner metrics to Apollo ([PR #5267](https://github.com/apollographql/router/pull/5267))

Reviewer 5: ```suggestion
### Send query planner metrics to Apollo ([PR #5267](https://github.com/apollographql/router/pull/5267))
```

>           otel.status_code:
>             static: error
>             condition:
>               eq:
>               - true
>               - on_graphql_error: true
>       supergraph:
>         attributes:
>           otel.status_code:
>             static: error
>             condition:
>               eq:
>               - true
>               - on_graphql_error: true
> ``` 
> 
> By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/5221
> 
> ### Instrument incremented on aborted request when condition is not fulfilled ([PR #5215](https://github.com/apollographql/router/pull/5215))
> 
> Previously when a telemetry instrument was dropped it would be incremented even if the associated condition was not fulfilled. For instance:
> 
> ```yaml
> telemetry:
>   instrumentation:
>     instruments:
>       router:
>         http.server.active_requests: false
>         http.server.request.duration: false
>         "custom_counter":
>           description: "count of requests"
>           type: counter
>           unit: "unit"
>           value: unit
>           # This instrument should not be triggered as the condition is never true
>           condition:
>             eq:
>               - response_header: "never-received"
>               - static: "true"
> ```
> 
> In the case where a request was started, but the client aborted the request before the response was sent, the `response_header` would never be set to `"never-received"`,
> and the instrument would not be triggered. However, the instrument would still be incremented.
> 
> Conditions are now checked for aborted requests, and the instrument is only incremented if the condition is fulfilled.
> 
> By [@BrynCooke](https://github.com/BrynCooke) in https://github.com/apollographql/router/pull/5215
> 
> ## 📃 Configuration
> 
> ### Send query planner metrics to Apollo ([PR #5267](https://github.com/apollographql/router/pull/5267))
> 
> To allow us to measure how much of an improvement the new query planner implementation makes, we are now transmitting metrics that start with:
> 
> `apollo.router.query_planning.*` to Apollo.
> 
> These metrics do not leak any sensitive information, but will greatly help us to improve query planning in the Router.
> 
> By [@BrynCooke](https://github.com/BrynCooke) in https://github.com/apollographql/router/pull/5267
> 

Author: ```suggestion
## 🐛 Fixes

### Fix custom attributes for spans and histogram when used with response_event ([PR #5221](https://github.com/apollographql/router/pull/5221))

It will fix several issues found:

+ Custom attributes based on response_event in spans were not added properly
+ Histograms using response_event selectors were not updated properly
+ Static selector to set a static value is now able to take a Value instead of just a string
+ Static selector to set a static value is now set at every stages
+ New `on_graphql_error` selector also available on supergraph
+ You can now override the status of a span using `otel.status_code` attribute to change the status of a span

For example, by default spans are marked in error if you have a critical error or http status code != 200, now if you want to mark your span in error if you have a graphql error in response body for example then you can have this kind of configuration:

```yaml
telemetry:
  instrumentation:
    spans:
      router:
        attributes:
          otel.status_code:
            static: error
            condition:
              eq:
              - true
              - on_graphql_error: true
      supergraph:
        attributes:
          otel.status_code:
            static: error
            condition:
              eq:
              - true
              - on_graphql_error: true
``` 

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/5221

### Instrument incremented on aborted request when condition is not fulfilled ([PR #5215](https://github.com/apollographql/router/pull/5215))

Previously when a telemetry instrument was dropped it would be incremented even if the associated condition was not fulfilled. For instance:

```yaml
telemetry:
  instrumentation:
    instruments:
      router:
        http.server.active_requests: false
        http.server.request.duration: false
        "custom_counter":
          description: "count of requests"
          type: counter
          unit: "unit"
          value: unit
          # This instrument should not be triggered as the condition is never true
          condition:
            eq:
              - response_header: "never-received"
              - static: "true"
```

In the case where a request was started, but the client aborted the request before the response was sent, the `response_header` would never be set to `"never-received"`,
and the instrument would not be triggered. However, the instrument would still be incremented.

Conditions are now checked for aborted requests, and the instrument is only incremented if the condition is fulfilled.

By [@BrynCooke](https://github.com/BrynCooke) in https://github.com/apollographql/router/pull/5215

## 🛠 Maintenance

### Send query planner and lifecycle metrics to Apollo ([PR #5267](https://github.com/apollographql/router/pull/5267), [PR #5270](https://github.com/apollographql/router/pull/5270))

To allow us to measure how much of an improvement the new query planner implementation makes, we are now transmitting to Apollo new metrics:

- `apollo.router.query_planning.*`  will greatly help us to improve query planning in the Router.
- `apollo.router.lifecycle.api_schema` provides us feedback on the experimental Rust-based API schema generation.
- `apollo.router.lifecycle.license` provides metrics on license expiration. We use this to improve the reliability of the license check mechanism.

These metrics do not leak any sensitive information.

By [@BrynCooke](https://github.com/BrynCooke) in https://github.com/apollographql/router/pull/5267, [@goto-bus-stop](https://github.com/goto-bus-stop)
```

## 🧪 Experimental

> ### Add graphql instruments ([PR #5215](https://github.com/apollographql/router/pull/5215), [PR #5257](https://github.com/apollographql/router/pull/5257))

Reviewer 5: ```suggestion
### Graphql instruments ([PR #5215](https://github.com/apollographql/router/pull/5215), [PR #5257](https://github.com/apollographql/router/pull/5257))
```

This project adheres to [Semantic Versioning v2.0.0](https://semver.org/spec/v2.0.0.html).

# [1.48.0] - 2024-05-29

## 🚀 Features

> ### Preview demand control ([PR #5317](https://github.com/apollographql/router/pull/5317))
> 
> > Note that Demand Control is in preview. Configuration and performance may change in future releases.
> Demand control is a feature that allows you to control the cost of operations in the router, potentially rejecting requests that are too expensive that could bring down the Router or subgraph services.
> 

Reviewer 2: Using our somewhat-standardized callout here, rather than an ad-hoc bit.  Also, the description of the feature doesn't need to be "quoted", it should be the main detail before the configuration offering.

```suggestion
### Demand control preview ([PR #5317](https://github.com/apollographql/router/pull/5317))

> ⚠️ This is a preview for an [Enterprise feature](https://www.apollographql.com/blog/platform/evaluating-apollo-router-understanding-free-and-open-vs-commercial-features/) of the Apollo Router. It requires an organization with a [GraphOS Enterprise plan](https://www.apollographql.com/pricing/).  If your organization doesn't currently have an Enterprise plan, you can test out this functionality with a [free Enterprise trial](https://studio.apollographql.com/signup?type=enterprise-trial).
>
> As a preview feature, it's subject to our [Preview launch stage](https://www.apollographql.com/docs/resources/product-launch-stages/#preview) expectations and configuration and performance may change in future releases.

Demand control allows you to control the cost of operations in the router, potentially rejecting requests that are too expensive that could bring down the Router or subgraphs.

```

Reviewer 2: I would like @chandrikas to take a last look at this.

Reviewer 3: Looks good to me!

```yaml
# Demand control enabled, but in measure mode.
preview_demand_control:
  enabled: true
  # `measure` or `enforce` mode. Measure mode will analyze cost of operations but not reject them.
  mode: measure

  strategy:
    # Static estimated strategy has a fixed cost for elements and when set to enforce will reject
    # requests that are estimated as too high before any execution takes place.
    static_estimated:
      # The assumed returned list size for operations. This should be set to the maximum number of items in graphql list
      list_size: 10
      # The maximum cost of a single operation.
      max: 1000
```

Telemetry is emitted for demand control, including the estimated cost of operations and whether they were rejected or not.
Full details will be included in the documentation for demand control which will be finalized before the next release.

By [@bryncooke](https://github.com/bryncooke) in https://github.com/apollographql/router/pull/5317

> ### Add support for studio trace id in selectors and document context entry ([Issue #3803](https://github.com/apollographql/router/issues/3803)), ([Issue #5172](https://github.com/apollographql/router/issues/5172))

Reviewer 2: ```suggestion
### Ability to include Apollo Studio trace ID on tracing spans ([Issue #3803](https://github.com/apollographql/router/issues/3803)), ([Issue #5172](https://github.com/apollographql/router/issues/5172))
```

Add support for a new trace ID selector kind, the `apollo` trace ID, which represents the trace ID on [Apollo GraphOS Studio](https://studio.apollographql.com/). 

An example configuration using `trace_id: apollo`:

```yaml
telemetry:
  instrumentation:
    spans:
      router:
        "studio.trace.id":

>             trace_id:: apollo

Reviewer 2: This looks wrong.

```suggestion
            trace_id: apollo
```

```

> Also, router documentation has been updated with constants that are available in Rhai.

Reviewer 2: @bnjjj This is really valuable!  But doesn't seem related to **this** changelog entry at all which makes it quite confusing to me. Can we just add a separate changelog entry that puts a "Documentation" update in?  It looks like it refers to:

https://github.com/apollographql/router/pull/5189/files#diff-5888944c765e2b49417fb75dd8c0e518bf4cb68fc264e6963f5be2da03503402R340-R352

Reviewer 2: After I submit this review, you'll see that I've made the suggestion below to fix this, therefore, I'll suggest removing this part, too (and update this comment with a link, afterwards).

Reviewer 7: Yes you're right, @Geal you can apply these changes

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/5189

### Add ability for router to deal with query plans with contextual rewrites ([PR #5097](https://github.com/apollographql/router/pull/5097))

Adds the ability for the router to execute query plans with context rewrites. A context is generated by the `@fromContext` directive, and each context maps values in the collected data JSON onto a variable that's used as an argument to a field resolver. To learn more, see [Saving and referencing data with contexts](https://www.apollographql.com/docs/federation/federated-types/federated-directives#saving-and-referencing-data-with-contexts).

⚠️ Because this feature requires a new version of federation, v2.8.0, distributed caches will need to be repopulated.

By [@clenfest](https://github.com/clenfest) in https://github.com/apollographql/router/pull/5097

## 🐛 Fixes

> ### Fix custom attributes for spans and histogram when used with response_event ([PR #5221](https://github.com/apollographql/router/pull/5221))

Reviewer 2: ```suggestion
### Fix custom attributes for spans and histogram when used with `response_event` ([PR #5221](https://github.com/apollographql/router/pull/5221))
```

Conditions are now checked for aborted requests, and the instrument is only incremented if the condition is fulfilled.

By [@BrynCooke](https://github.com/BrynCooke) in https://github.com/apollographql/router/pull/5215

## 🛠 Maintenance

### Send query planner and lifecycle metrics to Apollo ([PR #5267](https://github.com/apollographql/router/pull/5267), [PR #5270](https://github.com/apollographql/router/pull/5270))

To enable the performance measurement of the router's new query planner implementation, the router transmits to Apollo the following new metrics:
- `apollo.router.query_planning.*` provides metrics on the query planner that help improve the query planning implementation.
- `apollo.router.lifecycle.api_schema` provides feedback on the experimental Rust-based API schema generation.
- `apollo.router.lifecycle.license` provides metrics on license expiration that help improve the reliability of the license check mechanism.

These metrics don't leak any sensitive data.

By [@BrynCooke](https://github.com/BrynCooke) in https://github.com/apollographql/router/pull/5267, [@goto-bus-stop](https://github.com/goto-bus-stop)

> 

Reviewer 2: Completing my suggestion above elsewhere with its resolution. 😄 

```suggestion

## 📚 Documentation

### Add Rhai API constants reference

The Rhai API  documentation now includes [a list of available constants](https://www.apollographql.com/docs/router/customizations/rhai-api/#available-constants) that are available in the Rhai runtime.

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/5189
```
